src/Runner.Worker/StepsRunner.cs

```diff
@@ -352,7 +364,70 @@ private void CompleteStep(IStep step, TaskResult? result = null, string resultCo
         {
             var executionContext = step.ExecutionContext;
 
+            if (ShouldUploadAttachment(step))
+            {
+                QueueStepSummaryUpload(step);
+            }
+
             executionContext.Complete(result, resultCode: resultCode);
         }
+
+        private bool ShouldUploadAttachment(IStep step)
+        {
+          var executionContext = step.ExecutionContext;
+          var stepSummaryFilePath = executionContext.GetGitHubContext("step_summary");
+
+          if(!ActionCommandManager.StepSummaryEnabled(step.ExecutionContext))
+          {
+              Trace.Info("Step Summary is disabled; skipping attachment upload");
+              return false;
+          }
+
+          var summaryExists = File.Exists(stepSummaryFilePath);
+          if (summaryExists)
+          {
+              Trace.Info($"Summary file ({stepSummaryFilePath}) does not exist; skipping attachment upload");
+              return false;
+          }
+          Trace.Info($"Summary file exists: {stepSummaryFilePath}");
+
+          var summaryFileIsEmpty = new FileInfo(stepSummaryFilePath).Length == 0;
+          if (summaryFileIsEmpty)
+          {
+              Trace.Info($"Summary file ({stepSummaryFilePath}) is empty; skipping attachment upload");
+              return false;
+          }
+
+          return true;
+        }
+
+        private string CreateStepSummaryFile(IStep step)
+        {
+            var stepSummaryDirectory = HostContext.GetDirectory(WellKnownDirectory.StepSummary);
+            if (!Directory.Exists(stepSummaryDirectory))
+            {
+                Trace.Info($"Creating step summary directory: {stepSummaryDirectory}");
+                Directory.CreateDirectory(stepSummaryDirectory);
+            }
+
+            var stepID = step.ExecutionContext.Id;
+            var stepSummaryFilePath = Path.Combine(stepSummaryDirectory, $"{stepID}.md");
+
+            Trace.Info($"Creating step summary file: {stepSummaryFilePath}");
+            File.Create(stepSummaryFilePath).Close();
+            return stepSummaryFilePath;
+        }
+
+        private void QueueStepSummaryUpload(IStep step)
+        {
+            var executionContext = step.ExecutionContext;
+            var parentContext = executionContext.Root;
+            var stepSummaryFilePath = executionContext.GetGitHubContext("step_summary");
+            var stepID = step.ExecutionContext.Id;
+            var attachmentName = stepID.ToString();
```

Comment:
My understanding is we create the file, and the user can put arbitrary data into it because we provide them with the Path, is that correct?

If so, we will want to run the Secret Masker on this file here, and replace the contexts with the secret masked contents

Reply:
That could be done by getting all the lines in the file, then running them through the secret masker like

            foreach (string line in File.ReadLines())
            {
                var maskedLine = HostContext.SecretMasker.MaskSecrets(telemetryItem.Message);
                // rewrite the line here
            }

Reply:
We can probably move this secret scrubbing logic to when the file is dequeued and processed in [ServerQueue](https://github.com/actions/runner/blob/67ba8a7d429d953396c08550d8590af9889c4efe/src/Runner.Common/JobServerQueue.cs#L612):

ie:

```c#
else if (String.Equals(file.Type, ChecksAttachmentType.StepSummary, StringComparison.OrdinalIgnoreCase))
{
    // read step summary line by line
    // mask secrets & rewrite
    // send to CreateAttachmentAsync
}
```

cc @pfleidi

Reply:
That makes sense to me, but we want to be 100% sure we do this before merging :). If it fails in any way (cannot write to file), we should abort the upload

Reply:
I would suggest the opposite, which is don't change `JobServerQueue`.
- Log files that get queued to `JobServerQueue` are already run through the secret masker, I would suggest Attachment follow the same pattern.

Reply:
@TingluoHuang That makes sense. I already have some code in place locally to scrub secrets from the markdown files. It uses `SecretUtility.ScrubSecrets()` like this: ```
var scrubbedSummaryContent = SecretUtility.ScrubSecrets(stepSummaryContent, false);
```

I'll look into `SecretMasker`, though. Is there any preference of one vs. the other?

Reply:
Always use `HostContext.SecretMasker` since it has all the right secrets + encoder registered.

Reply:
@TingluoHuang 

> Log files that get queued to JobServerQueue are already run through the secret masker

Is this only when the type is `CoreAttachmentType.Log`? I'm having trouble finding the code reference, just wanted to make sure it hits the same codepath for `ChecksAttachmentType.StepSummary`.

Reply:
yes, `PagingLogger` is for regular step log.

Reply:
we have some bad hardcode string in there... 😢 
```
_jobServerQueue.QueueFileUpload(_timelineId, _timelineRecordId, "DistributedTask.Core.Log", "CustomToolLog", _dataFileName, true);
```